util/jsonrpc/jsonrpc.go

```diff
@@ -294,37 +289,63 @@ func (client *RPCClient) setConnectionError(err error) {
 // desktop applications, but we store the active connection and ping it regularly to
 // keep it alive (see ping()).
 func (client *RPCClient) conn() (*connection, error) {
-	if client.connection == nil {
-		defer client.connLock.Lock()()
-		if client.connection == nil {
-			defer client.backendsLock.RLock()()
-			start := 0
-			if len(client.backends) > 0 {
-				start = rand.Intn(len(client.backends))
-			}
-			var lastErr error
-			for i := 0; i < len(client.backends); i++ {
-				client.log.Debugf("Trying to connect to backend %v", client.backends[start].Name)
-				lastErr = client.establishConnection(client.backends[start])
-				if lastErr != nil {
-					client.log.WithError(lastErr).Info("Failover: backend is down")
-					start = (start + 1) % len(client.backends)
-				} else {
-					client.log.Debug("Successfully connected to backend")
-					break
-				}
-			}
-			if lastErr != nil {
-				client.log = client.log.WithField("backend", "offline")
-				// tried all backends
-				client.log.Debug("Disconnected from all backends")
-				client.setConnectionError(lastErr)
-				return nil, lastErr
+	client.connLock.RLock()
+	conn := client.connection
+	client.connLock.RUnlock()
+	if conn != nil {
+		return conn, nil
+	}
+
+	conn, err := func() (*connection, error) {
+		client.connLock.Lock()
+		defer client.connLock.Unlock()
+		if client.connection != nil {
+			return client.connection, nil
+		}
+
+		start := 0
+		if len(client.backends) > 0 {
```

Comment:
I think this would be pretty bad if `len(client.backens)` is zero, no? Then the `lastErr` below is always nil since the loop is never run and I wouldn't be surprised if this resulted in a nil dereference panic when accessing `client.connection` - it looks like it'll be `nil`.

Reply:
Good catch. Fixed